target/linux/x86/base-files/lib/upgrade/platform.sh

```diff
@@ -69,7 +112,54 @@ platform_do_bootloader_upgrade() {
 }
 
 platform_do_upgrade() {
-	local diskdev partdev diff
+	local _alt altboot bootpart defboot diff diskdev partdev partdevalt part size start
+
+	#extract the boot sector from the image
+	get_image "$@" | dd of=/tmp/image.bs count=63 bs=512b 2>/dev/null
+
+	if is_efi_system; then
+		export_partdevice_label bootpart openwrt_boot \
+			&& export_partdevice_label partdev openwrt_rootfs \
+			&& export_partdevice_label partdevalt openwrt_rootfs_alt || {
+				echo "Unable to determine upgrade device" >&2
+				return 1
+			}
+
+		grep -q 'root=PARTLABEL=openwrt_rootfs_alt' /proc/cmdline && _alt= || _alt=_alt
+		get_partitions /tmp/image.bs image
+		while read part start size; do
+			case "$part" in
+			1)
+				echo "Extracting bootloader from image..."
+				get_image "$@" | dd of=/tmp/.grub.img ibs="512" obs=1M skip="$start" count="$size" conv=fsync
+			;;
+			2)
+				echo "Writing new kernel to /boot/vmlinuz$_alt..."
+				if [ -n "$_alt" ]; then defboot=1; altboot=0; else defboot=0; altboot=1; fi
```

Comment:
See other comment.  `$_alt` shouldn't be true or false, but numeric.  This gets rid of all the `if` statements resulting in better consistency.